### PR TITLE
Limit length and content of account details

### PR DESCRIPTION
**Pull request by Parth097**
https://github.com/isaaccomputerscience/isaac-cs-issues/issues/450

### CODE REVIEW

**Comment by sxpydo** [comment]
The `Node.js CI / Build, Lint and Test (pull_request)` is failing in both FE and API

**Comment by Parth097** [comment]
> The `Node.js CI / Build, Lint and Test (pull_request)` is failing in both FE and API

Not sure why this is failing, I have not changed anything to do with this.

File: config-templates/content_indices.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
 #
 #Mon Jan 29 14:17:14 BST 2023
-latest=05bbbb74fd52cd0aea3a048a3c1e6c434ce5efaf
-live=05bbbb74fd52cd0aea3a048a3c1e6c434ce5efaf
+latest=4bde63432762588c3e83ecdbd74485dfe23b2885
+live=4bde63432762588c3e83ecdbd74485dfe23b2885
```
Comment (hannah-whelan) [inline config-templates/content_indices.properties:4]: I don't think a content SHA update is required for this ticket?

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java [hunk 2]
```diff
@@ -125,9 +126,10 @@ private static Stream<Arguments> namesToTest() {
         Arguments.of(false, null), // Null is not valid
         Arguments.of(false, ""), // Empty string is not valid
         Arguments.of(false, " "), // Only whitespace is not valid
-        Arguments.of(false, "a".repeat(256)), // Names exceeding the maximum length of 255 characters are not valid
+        Arguments.of(true, "a".repeat(50)), // The maximum length of 50 characters are valid
+        Arguments.of(false, "a".repeat(51)), // Names exceeding the maximum length of 255 characters are not valid
```
Comment (hannah-whelan) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java:130]: Comment here still has previous 255 length

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java [hunk 3]
```diff
@@ -149,10 +149,9 @@ public class UserAccountManager implements IUserAccountManager {
   private final SchoolListReader schoolListReader;
 
   private final Pattern restrictedSignupEmailRegex;
-  private static final int USER_NAME_MAX_LENGTH = 255;
-  private static final Pattern USER_NAME_PERMITTED_CHARS_REGEX =
-      Pattern.compile("^[\\p{L}\\d_\\-' ]+$", Pattern.CANON_EQ);
+  private static final int USER_NAME_MAX_LENGTH = 50;
   private static final Pattern EMAIL_PERMITTED_CHARS_REGEX = Pattern.compile("^[a-zA-Z0-9!#$%&'+\\-=?^_`.{|}~@]+$");
+  private static final Pattern USER_NAME_PERMITTED_CHARS_REGEX = Pattern.compile("^[\\p{L}\\-'`! ]+$");
```
Comment (hannah-whelan) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java:154]: This new regex now allows numbers to be used in user names (first / last name) - why? Backticks and exclamation marks are also now permitted, where they weren't before; they shouldn't be needed in a name?

It has also removed the ability to use an underscore (though I'm assuming that one is intentional?)

File: config-templates/content_indices.properties (hunk 1 above)
Comment (Parth097) [reply config-templates/content_indices.properties:4]: This is not for the ticket, just to have updated events and such things. Having this wouldn't hurt for the future.
If you still feel like we dont want new SHA then i'll revert it.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java [hunk 4]
```diff
@@ -114,8 +114,9 @@ private static Stream<Arguments> namesToTest() {
         Arguments.of(true, "testname"), // Lowercase is valid
         Arguments.of(true, "TESTNAME"), // Uppercase is valid
         Arguments.of(true, "TestName"), // Mixture of cases is valid
-        Arguments.of(true, "TestName123"), // Numbers are valid
-        Arguments.of(true, "Test_Name"), // Underscores are valid
+        Arguments.of(true, "TestName`"), // Mixture of cases is valid
+        Arguments.of(false, "TestName123"), // Numbers are valid
+        Arguments.of(false, "Test_Name"), // Underscores are valid
```
Comment (hannah-whelan) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java:119]: underscores are now invalid

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java [hunk 5]
```diff
@@ -125,9 +126,10 @@ private static Stream<Arguments> namesToTest() {
         Arguments.of(false, null), // Null is not valid
         Arguments.of(false, ""), // Empty string is not valid
         Arguments.of(false, " "), // Only whitespace is not valid
-        Arguments.of(false, "a".repeat(256)), // Names exceeding the maximum length of 255 characters are not valid
+        Arguments.of(true, "a".repeat(50)), // The maximum length of 50 characters are valid
+        Arguments.of(false, "a".repeat(51)), // Names exceeding the maximum length of 255 characters are not valid
         // Other special characters are not permitted
-        Arguments.of(false, "Test!Name"),
+        Arguments.of(true, "Test!Name"),
```
Comment (hannah-whelan) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java:132]: This is now permitting exclamation marks

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java [hunk 6]
```diff
@@ -114,8 +114,9 @@ private static Stream<Arguments> namesToTest() {
         Arguments.of(true, "testname"), // Lowercase is valid
         Arguments.of(true, "TESTNAME"), // Uppercase is valid
         Arguments.of(true, "TestName"), // Mixture of cases is valid
-        Arguments.of(true, "TestName123"), // Numbers are valid
-        Arguments.of(true, "Test_Name"), // Underscores are valid
+        Arguments.of(true, "TestName`"), // Mixture of cases is valid
+        Arguments.of(false, "TestName123"), // Numbers are valid
```
Comment (hannah-whelan) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java:118]: now invalid

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java (hunk 2 above)
Comment (Parth097) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java]: changed comment

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java (hunk 4 above)
Comment (Parth097) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java]: changed comment

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java (hunk 5 above)
Comment (Parth097) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java]: changed the location of the test. It is now above the comment.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java (hunk 6 above)
Comment (Parth097) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java]: changed comment

File: config-templates/content_indices.properties (hunk 1 above)
Comment (hannah-whelan) [reply config-templates/content_indices.properties:4]: We usually don't update it in a ticket unless it's required (such as adding new email template required by the code for example)

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java (hunk 3 above)
Comment (Parth097) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java]: Back ticks are invalid now, ! & _ are allowed

File: config-templates/content_indices.properties (hunk 1 above)
Comment (Parth097) [reply config-templates/content_indices.properties]: reverted back

**Review by hannah-whelan — APPROVED** [review]

**Pull request merged**